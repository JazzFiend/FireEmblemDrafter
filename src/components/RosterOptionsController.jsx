import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';

import Dropdown from './Dropdown/Dropdown';

class RosterOptionsController extends PureComponent {
  static displayList(list, label) {
    if (!list.length) {
      return '';
    }
    let listDisplayFriendly = label;

    _.forEach(list, (value) => {
      listDisplayFriendly = `${listDisplayFriendly + value}\n`;
    });
    return listDisplayFriendly;
  }

  static _keyList(list) {
    const keyedList = [];
    _.forEach(list, (element, index) => {
      keyedList.push(
        {
          id: index,
          title: element,
        },
      );
    });
    return keyedList;
  }

  static filterRoster(allCharacters, restrictedCharacters, requiredCharacters) {
    return allCharacters.filter(
      (character) => (!restrictedCharacters.includes(character)),
    ).filter(
      (character) => (!requiredCharacters.includes(character)),
    );
  }

  // TODO: These two functions are nearly identical, but its not a typical refactor. I think I need to create a
  // UnitDropdown class or something. Going to save this for its own task.
  showRequiredDropdown(currentRosterKeyed) {
    const { draftInProgress, allCharacters, handleRequiredUnitSelector } = this.props;
    return (
      <div data-testid="required-drop-down">
        {
      (!draftInProgress && allCharacters.length > 0)
      && (
      <Dropdown
        defaultText="Add Required..."
        dropdownItems={currentRosterKeyed}
        onClick={(characterId) => handleRequiredUnitSelector(currentRosterKeyed.find(
          (element) => element.id === characterId,
        ).title)}
      />
      )
    }
      </div>
    );
  }

  showRestrictedDropdown(currentRosterKeyed) {
    const { draftInProgress, allCharacters, handleRestrictedUnitSelector } = this.props;
    return (
      <div data-testid="restricted-drop-down">
        {
      (!draftInProgress && allCharacters.length > 0)
      && (
      <Dropdown
        defaultText="Add Restricted..."
        dropdownItems={currentRosterKeyed}
        onClick={(characterId) => handleRestrictedUnitSelector(currentRosterKeyed.find(
          (element) => element.id === characterId,
        ).title)}
      />
      )
    }
      </div>
    );
  }

  render() {
    const { restrictedCharacters, requiredCharacters, allCharacters } = this.props;
    const filteredRoster = RosterOptionsController.filterRoster(
      allCharacters,
      restrictedCharacters,
      requiredCharacters,
    );
    const currentRosterKeyed = RosterOptionsController._keyList(filteredRoster);

    return (
      <div>
        {this.showRequiredDropdown(currentRosterKeyed)}
        <div data-testid="required-list">
          {RosterOptionsController.displayList(requiredCharacters, 'Required Units: ')}
        </div>
        {this.showRestrictedDropdown(currentRosterKeyed)}
        <div data-testid="restricted-list">
          {RosterOptionsController.displayList(restrictedCharacters, 'Restricted Units: ')}
        </div>
      </div>
    );
  }
}

RosterOptionsController.propTypes = {
  restrictedCharacters: PropTypes.arrayOf(PropTypes.string).isRequired,
  requiredCharacters: PropTypes.arrayOf(PropTypes.string).isRequired,
  allCharacters: PropTypes.arrayOf(PropTypes.string),
  handleRestrictedUnitSelector: PropTypes.func.isRequired,
  handleRequiredUnitSelector: PropTypes.func.isRequired,
  draftInProgress: PropTypes.bool.isRequired,
};

RosterOptionsController.defaultProps = {
  allCharacters: [],
};

export default RosterOptionsController;
