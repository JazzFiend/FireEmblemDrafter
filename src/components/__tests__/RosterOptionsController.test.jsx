import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import userEvent from '@testing-library/user-event';

import RosterOptionsController from '../RosterOptionsController';

test('should render nothing when draft has not started false and no characters are loaded', () => {
  const { asFragment } = render(
    <RosterOptionsController
      restrictedCharacters={[]}
      requiredCharacters={[]}
      allCharacters={[]}
      handleRestrictedUnitSelector={jest.fn()}
      handleRequiredUnitSelector={jest.fn()}
      draftInProgress={false}
    />,
  );
  expect(asFragment()).toMatchSnapshot();
});

test('should render unit dropdowns when draft has not started false and characters are loaded', () => {
  const { asFragment } = render(
    <RosterOptionsController
      restrictedCharacters={[]}
      requiredCharacters={[]}
      allCharacters={['u', 'v', 'w', 'x', 'y', 'z']}
      handleRestrictedUnitSelector={jest.fn()}
      handleRequiredUnitSelector={jest.fn()}
      draftInProgress={false}
    />,
  );
  expect(asFragment()).toMatchSnapshot();
});

// TODO: I think that all of these tests could be replaced by snapshot tests. But I want to try to simplify the
// Dropdown Component first.
test('should not render when a set of characters is not loaded', () => {
  const { getByTestId } = render(
    <RosterOptionsController
      restrictedCharacters={[]}
      requiredCharacters={[]}
      allCharacters={[]}
      handleRestrictedUnitSelector={() => {}}
      handleRequiredUnitSelector={() => {}}
      draftInProgress={false}
    />,
  );

  const restrictedDropdown = getByTestId('restricted-drop-down');
  expect(restrictedDropdown).toHaveTextContent('');
});

test('should render when a set of characters is loaded', () => {
  const allCharacters = ['u', 'v', 'w', 'x', 'y', 'z'];

  const { getByTestId } = render(
    <RosterOptionsController
      restrictedCharacters={[]}
      requiredCharacters={[]}
      allCharacters={allCharacters}
      handleRestrictedUnitSelector={() => {}}
      handleRequiredUnitSelector={() => {}}
      draftInProgress={false}
    />,
  );

  const restrictedDropdown = getByTestId('restricted-drop-down');
  expect(restrictedDropdown).toHaveTextContent('Restricted Units');
});

test('should hide dropdown and show restricted units when draft has started', () => {
  const restrictedCharacters = ['y', 'z'];
  const allCharacters = ['u', 'v', 'w', 'x', 'y', 'z'];

  const { getByTestId } = render(
    <RosterOptionsController
      restrictedCharacters={restrictedCharacters}
      requiredCharacters={[]}
      allCharacters={allCharacters}
      handleRestrictedUnitSelector={() => {}}
      handleRequiredUnitSelector={() => {}}
      draftInProgress={true}
    />,
  );

  const restrictedDropdown = getByTestId('restricted-drop-down');
  expect(restrictedDropdown).toHaveTextContent('');
  const restrictedList = getByTestId('restricted-list');
  expect(restrictedList).toHaveTextContent('Restricted Units: y z');
});

test('should hide dropdown and show required units when draft has started', () => {
  const requiredCharacters = ['y', 'z'];
  const allCharacters = ['u', 'v', 'w', 'x', 'y', 'z'];

  const { getByTestId } = render(
    <RosterOptionsController
      restrictedCharacters={[]}
      requiredCharacters={requiredCharacters}
      allCharacters={allCharacters}
      handleRestrictedUnitSelector={() => {}}
      handleRequiredUnitSelector={() => {}}
      draftInProgress={true}
    />,
  );

  const requiredDropdown = getByTestId('required-drop-down');
  expect(requiredDropdown).toHaveTextContent('');
  const requiredList = getByTestId('required-list');
  expect(requiredList).toHaveTextContent('Required Units: y z');
});

test('should run restrictedUnitSelector function when clicked', () => {
  const allCharacters = ['u', 'v', 'w', 'x', 'y', 'z'];
  const testClickHandler = jest.fn();
  const { container } = render(
    <RosterOptionsController
      restrictedCharacters={[]}
      requiredCharacters={[]}
      allCharacters={allCharacters}
      handleRestrictedUnitSelector={(id) => testClickHandler(id)}
      handleRequiredUnitSelector={() => {}}
      draftInProgress={false}
    />,
  );

  userEvent.click(container.querySelector('.restricted-selector__control'));
  userEvent.click(container.querySelectorAll('.restricted-selector__option')[0]);
  expect(testClickHandler).toHaveBeenCalledWith(['u']);
});

test('should run requiredUnitSelector function when clicked', () => {
  const allCharacters = ['u', 'v', 'w', 'x', 'y', 'z'];
  const testClickHandler = jest.fn();
  const { container } = render(
    <RosterOptionsController
      restrictedCharacters={[]}
      requiredCharacters={[]}
      allCharacters={allCharacters}
      handleRestrictedUnitSelector={() => {}}
      handleRequiredUnitSelector={(id) => testClickHandler(id)}
      draftInProgress={false}
    />,
  );

  userEvent.click(container.querySelector('.required-selector__control'));
  userEvent.click(container.querySelectorAll('.required-selector__option')[0]);
  expect(testClickHandler).toHaveBeenCalledWith(['u']);
});

test('options should disappear and become disabled in the required menu', () => {
  const allCharacters = ['u', 'v', 'w', 'x', 'y', 'z'];
  const { asFragment, container } = render(
    <RosterOptionsController
      restrictedCharacters={['u', 'v']}
      requiredCharacters={['w']}
      allCharacters={allCharacters}
      handleRestrictedUnitSelector={() => {}}
      handleRequiredUnitSelector={() => {}}
      draftInProgress={false}
    />,
  );

  userEvent.click(container.querySelector('.required-selector__control'));

  expect(asFragment()).toMatchSnapshot();
});
