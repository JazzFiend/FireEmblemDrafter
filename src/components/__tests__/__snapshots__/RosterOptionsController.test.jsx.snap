// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render nothing when draft has not started false and no characters are loaded 1`] = `
<DocumentFragment>
  <div>
    <div
      data-testid="required-drop-down"
    />
    <div
      data-testid="required-list"
    />
    <div
      data-testid="restricted-drop-down"
    />
    <div
      data-testid="restricted-list"
    />
  </div>
</DocumentFragment>
`;

exports[`should render unit dropdowns when draft has not started false and characters are loaded 1`] = `
<DocumentFragment>
  <div>
    <div
      data-testid="required-drop-down"
    >
      <div
        class="dd-wrapper"
      >
        <div
          class="dd-header"
          role="listbox"
          tabindex="0"
        >
          <div
            class="dd-header-title"
            data-testid="dropdown"
          >
            Add Required...
          </div>
        </div>
      </div>
    </div>
    <div
      data-testid="required-list"
    />
    <div
      data-testid="restricted-drop-down"
    >
      <div
        class="dd-wrapper"
      >
        <div
          class="dd-header"
          role="listbox"
          tabindex="0"
        >
          <div
            class="dd-header-title"
            data-testid="dropdown"
          >
            Add Restricted...
          </div>
        </div>
      </div>
    </div>
    <div
      data-testid="restricted-list"
    />
  </div>
</DocumentFragment>
`;
